src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1425,6 +1425,76 @@
       "ALTER TABLE environments RENAME COLUMN name TO environment"
       "ALTER TABLE environments ALTER COLUMN name RENAME TO environment")))
 
+(defn move-to-jsonb-for-metrics-logs-resources
+  []
+  (let [hash-type (if (sutils/postgres?) "bytea" "varchar(40)")
+        uuid-type (if (sutils/postgres?) "uuid" "varchar(255)")
+        jsonb-type (if (sutils/postgres?) "jsonb" "text")]
+    (sql/create-table :reports_transform
+                      ["id"                    "bigint NOT NULL DEFAULT nextval('reports_id_seq')"]
+                      ["hash"                  hash-type "NOT NULL"]
+                      ["transaction_uuid"      uuid-type]
+                      ["certname"              "text NOT NULL"]
+                      ["puppet_version"        "varchar(255) NOT NULL"]
+                      ["report_format"         "smallint NOT NULL"]
+                      ["configuration_version" "varchar(255) NOT NULL"]
+                      ["start_time"            "timestamp with time zone NOT NULL"]
+                      ["end_time"              "timestamp with time zone NOT NULL"]
+                      ["receive_time"          "timestamp with time zone NOT NULL"]
+                      ["producer_timestamp"    "timestamp with time zone NOT NULL"]
+                      ["noop"                  "boolean"]
+                      ["environment_id"        "bigint"]
+                      ["status_id"             "bigint"]
+                      ;; Add a `resources' column and change from "json" to
+                      ;; "jsonb" for the `logs' and `metrics' columns. Keep the
+                      ;; old `json' columns around to avoid painful migrations.
+                      ["resources" jsonb-type]
+                      ["metrics" jsonb-type]
+                      ["logs" jsonb-type])
```

Comment:
@wkalt pinged me that he was seeing negligible times with the direct migration from JSON to JSONB here, and I saw the same for 500 nodes/14 days of reports with the benchmarking tool, the whole migration was pretty fast, which makes sense given what I've read about the time to migrate growing with the number of fields in the JSON and how many fields there are in logs/metrics compared to a `full_report`.